src/main.rs

```diff
@@ -0,0 +1,464 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate byte_unit;
+#[macro_use]
+extern crate clap;
+#[macro_use]
+extern crate error_chain;
+extern crate flate2;
+#[macro_use]
+extern crate nix;
+extern crate progress_streams;
+extern crate reqwest;
+extern crate tempdir;
+extern crate xz2;
+
+mod blockdev;
+mod errors;
+mod source;
+mod verify;
+
+use byte_unit::Byte;
+use clap::{App, AppSettings, Arg};
+use error_chain::ChainedError;
+use flate2::read::GzDecoder;
+use progress_streams::ProgressReader;
+use reqwest::Url;
+use std::fs::{create_dir_all, read_dir, File, OpenOptions};
+use std::io::{copy, BufRead, BufReader, Read, Seek, SeekFrom, Write};
+use std::path::Path;
+use std::time::{Duration, Instant};
+use xz2::read::XzDecoder;
+
+use crate::blockdev::*;
+use crate::errors::*;
+use crate::source::*;
+use crate::verify::*;
+
+struct Config {
+    device: String,
+    location: Box<dyn ImageLocation>,
+    ignition: Option<String>,
+    platform: Option<String>,
+    firstboot_kargs: Option<String>,
+    insecure: bool,
+}
+
+quick_main!(run);
+
+fn run() -> Result<()> {
+    // get arguments
+    let config = parse_args().chain_err(|| "parsing arguments")?;
+
+    // set up image source
+    let mut source = config.location.source()?;
+    if source.signature.is_empty() {
+        if config.insecure {
+            eprintln!("Signature not found; skipping verification as requested");
+        } else {
+            bail!("--insecure not specified and signature not found");
+        }
+    }
+
+    // open output; ensure we have exclusive access
+    let mut dest = OpenOptions::new()
+        .write(true)
+        .open(&config.device)
+        .chain_err(|| format!("opening {}", &config.device))?;
+    reread_partition_table(&dest)
+        .chain_err(|| format!("checking for exclusive access to {}", &config.device))?;
+
+    // copy and postprocess disk image
+    // On failure, clear and reread the partition table to prevent the disk
+    // from accidentally being used.
+    let result = write_disk(&config, &mut source, &mut dest);
+    if result.is_err() {
+        // log the error so the details aren't dropped if we encounter
+        // another error during cleanup
+        eprint!("{}", ChainedError::display_chain(&result.unwrap_err()));
+
+        // clean up
+        clear_partition_table(&mut dest)?;
+
+        // return a generic error so our exit status is right
+        bail!("install failed");
+    }
+
+    println!("Install complete.");
+    Ok(())
+}
+
+// Parse command-line arguments.
```

Comment:
Here and pretty much everywhere else in the codebase: docstrings require three slashes `///`.